internal/fwschemadata/data_set.go

```diff
@@ -10,11 +10,31 @@ import (
 	"github.com/hashicorp/terraform-plugin-framework/diag"
 	"github.com/hashicorp/terraform-plugin-framework/internal/reflect"
 	"github.com/hashicorp/terraform-plugin-framework/path"
+	"github.com/hashicorp/terraform-plugin-go/tftypes"
 )
 
-// Set replaces the entire value. The value should be a struct whose fields
+// Set replaces the entire value. The value can be a tftypes.Value or a struct whose fields
 // have one of the attr.Value types. Each field must have the tfsdk field tag.
 func (d *Data) Set(ctx context.Context, val any) diag.Diagnostics {
+	var diags diag.Diagnostics
+
+	if v, ok := val.(tftypes.Value); ok {
+		objType := d.Schema.Type().TerraformType(ctx)
+
+		if !objType.Equal(v.Type()) {
+			diags.AddError(
+				d.Description.Title()+" Write Error",
+				"An unexpected error was encountered trying to write the "+d.Description.String()+". This is always an error in the provider. Please report the following to the provider developer:\n\n"+
+					fmt.Sprintf("Error: Type mismatch between provided value and type of %s", d.Description.String()),
```

Comment:
Might be useful to print out the expected type (`objType`) and the type we got (`v.Type()`) to make that easier to debug

Reply:
I umm'd and ahh'd over whether to print out the expected and received types for `Set`, particularly for schemas that have a lot of attributes + nesting I find the output of that practically unreadable/unusable unless it's manually manipulated into a more human readable format. 

But also maybe it's just worth providing the info upfront and side-step having to retrieve it some other way..